### PR TITLE
feat(superblocks): create superblocks even if there are no blocks during superepoch

### LINKED ISSUE #1415: Produce superblocks even if superepoch had no consolidated blocks

### PR DESCRIPTION
**Pull request by mariocao**
This PR creates superblocks even when no blocks have been proposed during a superblock period.
It also stores the last ARS instead of the prior to last ARS

Closes #1415

### CODE REVIEW

File: data_structures/src/superblock.rs [hunk 1]
```diff
@@ -175,63 +175,52 @@ impl SuperBlockState {
...
-                self.received_superblocks = old_superblock_votes;
+        );
 
-                Some(superblock)
+        self.current_superblock_hash = superblock.hash();
+
+        // Save ARS identities:
+        // previous = current
+        // current = ars_pkh_keys
+
```
Comment (aesedepece) [inline data_structures/src/superblock.rs:196]:
```suggestion

```

File: data_structures/src/superblock.rs [hunk 2]
```diff
@@ -265,31 +254,50 @@ impl SuperBlockState {
 }
 
 /// Produces a `SuperBlock` that includes the blocks in `block_headers` if there is at least one of them.
+/// // remove return Option
```
Comment (aesedepece) [inline data_structures/src/superblock.rs:257]:
```suggestion
```

File: data_structures/src/superblock.rs [hunk 3]
```diff
@@ -265,31 +254,50 @@ impl SuperBlockState {
...
+        {
+            SuperBlock {
+                ars_length: ars_ordered_hash_leaves.len() as u64,
+                ars_root: hash_merkle_tree_root(ars_ordered_hash_leaves),
+                data_request_root: Hash::default(),
+                tally_root: Hash::default(),
+                index,
+                last_block: last_block_in_previous_superblock,
+                last_block_in_previous_superblock,
+            }
```
Comment (aesedepece) [inline data_structures/src/superblock.rs:277]: I sort of like this better just to keep focus on the fields that are not being defaulted and are relevant to each case:
```suggestion
            SuperBlock {
                ars_length: ars_ordered_hash_leaves.len() as u64,
                ars_root: hash_merkle_tree_root(ars_ordered_hash_leaves),
                index,
                last_block: last_block_in_previous_superblock,
                last_block_in_previous_superblock,
                ..Default::default()
            }
```

Note however that this is a code style preference, I have no real problem with how it is right now :sweat_smile:

File: data_structures/src/superblock.rs [hunk 4]
```diff
@@ -265,31 +254,50 @@ impl SuperBlockState {
...
+
+            SuperBlock {
+                ars_length: ars_ordered_hash_leaves.len() as u64,
+                data_request_root: hash_merkle_tree_root(&merkle_drs),
+                tally_root: hash_merkle_tree_root(&merkle_tallies),
+                ars_root: hash_merkle_tree_root(ars_ordered_hash_leaves),
+                index,
+                last_block: last_block_hash,
+                last_block_in_previous_superblock,
+            }
```
Comment (aesedepece) [inline data_structures/src/superblock.rs:298]: Order fields alphabetically or @Tommytrg's ghost will haunt you at night:
```suggestion
            SuperBlock {
                ars_length: ars_ordered_hash_leaves.len() as u64,
                ars_root: hash_merkle_tree_root(ars_ordered_hash_leaves),
                data_request_root: hash_merkle_tree_root(&merkle_drs),
                index,
                last_block: last_block_hash,
                last_block_in_previous_superblock,
                tally_root: hash_merkle_tree_root(&merkle_tallies),
            }
```

File: data_structures/src/superblock.rs [hunk 5]
```diff
@@ -515,12 +528,43 @@ mod tests {
...
 
-        let mut expected_sbs = SuperBlockState::default();
+        let expected_superblock = SuperBlock {
+            ars_length: 1,
+            data_request_root: Hash::default(),
+            tally_root: Hash::default(),
+            ars_root: hash_merkle_tree_root(&hash_key_leaves(&[bls_pk.clone()])),
+            index: 0,
+            last_block: genesis_hash,
+            last_block_in_previous_superblock: genesis_hash,
```
Comment (aesedepece) [inline data_structures/src/superblock.rs:547]: Order!

File: data_structures/src/superblock.rs [hunk 6]
```diff
@@ -552,23 +561,20 @@ mod tests {
...
+        );
         let v1 = SuperBlockVote::new_unsigned(sb1.hash(), 0);
         assert_eq!(sbs.add_vote(&v1), AddSuperBlockVote::NotInSigningCommittee);
     }
 
     #[test]
-    fn superblock_state_first_superblock_none() {
+    fn superblock_state_first_superblock_empty() {
         // If the first superblock is None, the state is not updated except for the superblock_index
         let mut sbs = SuperBlockState::default();
```
Comment (clbartoli) [inline data_structures/src/superblock.rs:580]: i think the comment below should be updated to "If there were no blocks, there will be an empty superblock"

Comment (mariocao) [reply data_structures/src/superblock.rs:580]: True... it got lost during the rebase... :rofl:

File: data_structures/src/superblock.rs [hunk 7]
```diff
@@ -327,26 +312,43 @@ pub fn mining_build_superblock(
...
+        None =>
+        // Build "empty" Superblock (there were no blocks during super-epoch)
+        {
+            SuperBlock {
+                ars_length: ars_ordered_hash_leaves.len() as u64,
+                ars_root: hash_merkle_tree_root(ars_ordered_hash_leaves),
+                index,
+                last_block: last_block_in_previous_superblock,
+                last_block_in_previous_superblock,
+                ..Default::default()
```
Comment (tmpolaczyk) [inline data_structures/src/superblock.rs:327]: Using `Default::default()` here sets the remaining fields to `Hash::SHA256([0; 32])`. Currently the hash of an empty merkle tree is set to the empty hash `e3b0c44298f...`. It sure makes sense to use `00000000...` as the default hash, but we should be consistent.

**Comment by lrubiorod** [comment]
During synchronizing I saw several messages of GetBlockEpochRange, it is normal?
image

File: data_structures/src/superblock.rs (hunk 7 above)
Comment (mariocao) [reply data_structures/src/superblock.rs:327]: Agreed! It was seen but not changed...

**Pull request merged**